request.go

```diff
@@ -582,6 +589,13 @@ func filestat(h FileLister, r *Request, pkt requestPacket) responsePacket {
 			return statusFromError(pkt.id(), err)
 		}
 		filename := finfo[0].Name()
+		if rl, ok := (finfo[0]).(LinkReader); ok {
+			if f, err := rl.ReadLink(); err != nil {
+				return statusFromError(pkt.id(), err)
+			} else {
+				filename = f
+			}
```

Comment:
This is better written as:
```
			f, err := rl.ReadLink()
			if err != nil {
				return statusFromError(pkt.id(), err)
			}
			filename = f
```

Reply:
> We could do something like this:
> 
> ```
> diff --git a/request-interfaces.go b/request-interfaces.go
> index 2e5ee6b..ab5f5a1 100644
> --- a/request-interfaces.go
> +++ b/request-interfaces.go
> @@ -94,7 +94,7 @@ type LstatFileLister interface {
>  //
>  // Up to v1.13.5 the signature for the RealPath method was:
>  //
> -// RealPath(string) string
> +// # RealPath(string) string
>  //
>  // we have added a legacyRealPathFileLister that implements the old method
>  // to ensure that your code does not break.
> @@ -104,6 +104,13 @@ type RealPathFileLister interface {
>         RealPath(string) (string, error)
>  }
>  
> +// ReadlinkFileLister is a FileLister that implements the Readlink method.
> +// TODO: complete docs
> +type ReadlinkFileLister interface {
> +       FileLister
> +       Readlink(string) (string, error)
> +}
> +
>  // This interface is here for backward compatibility only
>  type legacyRealPathFileLister interface {
>         FileLister
> diff --git a/request.go b/request.go
> index 6a7e6cf..4db3f54 100644
> --- a/request.go
> +++ b/request.go
> @@ -295,7 +295,25 @@ func (r *Request) call(handlers Handlers, pkt requestPacket, alloc *allocator, o
>                 return filecmd(handlers.FileCmd, r, pkt)
>         case "List":
>                 return filelist(handlers.FileList, r, pkt)
> -       case "Stat", "Lstat", "Readlink":
> +       case "Stat", "Lstat":
> +               return filestat(handlers.FileList, r, pkt)
> +       case "Readlink":
> +               if readlinkFileLister, ok := handlers.FileList.(ReadlinkFileLister); ok {
> +                       resolved, err := readlinkFileLister.Readlink(r.Filepath)
> +                       if err != nil {
> +                               return statusFromError(pkt.id(), err)
> +                       }
> +                       return &sshFxpNamePacket{
> +                               ID: pkt.id(),
> +                               NameAttrs: []*sshFxpNameAttr{
> +                                       {
> +                                               Name:     resolved,
> +                                               LongName: resolved,
> +                                               Attrs:    emptyFileStat,
> +                                       },
> +                               },
> +                       }
> +               }
>                 return filestat(handlers.FileList, r, pkt)
>         default:
>                 return statusFromError(pkt.id(), fmt.Errorf("unexpected method: %s", r.Method))
> ```
> 
> @puellanivis, @georgmu do you have other/better suggestions? thanks

I think this sounds like a good solution and is done in the same way as the .

Another option would be a bigger rewrite but this would brake existing code: Make `Handlers` an Interface (it is used way in most cases anyway where FileGet/FilePut/FileCmd/FileList all point to the same implementation) and make the now optional interfaces mandatory with a hint on what to return when support is not necessary. Go is lacking features like default implementations for interfaces like in rust...

Reply:
> Another option would be a bigger rewrite but this would brake existing code: Make Handlers an Interface (it is used way in most cases anyway where FileGet/FilePut/FileCmd/FileList all point to the same implementation) and make the now optional interfaces mandatory with a hint on what to return when support is not necessary. Go is lacking features like default implementations for interfaces like in rust...

This will require a jump to V2. But this is something that we really do need to get done. Default implementations can be done by allowing embedded types that implement the default. See protobuf and their “UnimplementedXYServer”.

We can also cover default implementations through interface extensions as we have been doing a few of the newer implementations.

I understand that it is hard to shift out of a familiar paradigm where rust provides for default implementations, but there exists an equivalent in Go. You just have to often think about things from a different perspective.